diff --recursive '--context=5' slurm-bank-1.4.2/slurm-bank.spec slurm-bank-1.4.2-new/slurm-bank.spec
*** slurm-bank-1.4.2/slurm-bank.spec	2022-08-20 15:11:48.000000000 -0700
--- slurm-bank-1.4.2-new/slurm-bank.spec	2023-02-16 10:10:02.003434446 -0800
***************
*** 1,10 ****
  # Pass --with docs to rpmbuild if you want the HTML documentation (requires ikiwiki)
  
  Name:           slurm-bank
  Version:        1.4.2
! Release:        1%{?dist}
  Summary:        SLURM Bank, a collection of wrapper scripts to do banking
  
  Group:          System
  License:        GPLv2
  URL:            http://www.tchpc.tcd.ie/
--- 1,10 ----
  # Pass --with docs to rpmbuild if you want the HTML documentation (requires ikiwiki)
  
  Name:           slurm-bank
  Version:        1.4.2
! Release:        2%{?dist}
  Summary:        SLURM Bank, a collection of wrapper scripts to do banking
  
  Group:          System
  License:        GPLv2
  URL:            http://www.tchpc.tcd.ie/
diff --recursive '--context=5' slurm-bank-1.4.2/src/_sbank-balance.pl slurm-bank-1.4.2-new/src/_sbank-balance.pl
*** slurm-bank-1.4.2/src/_sbank-balance.pl	2022-08-20 15:03:51.000000000 -0700
--- slurm-bank-1.4.2-new/src/_sbank-balance.pl	2023-02-16 10:11:24.930151518 -0800
***************
*** 91,101 ****
  		printf "%-10s %9s | %14s %9s\n",
  			"User", "Usage", "Account", "Usage";
  		printf "%10s %9s + %14s %9s\n",
  			"-"x10, "-"x9, "-"x14, "-"x9;
  	} else {
! 		printf "%-10s %9s | %14s %9s | %13s %9s (CPU hrs)\n",
  			"User", "Usage", "Account", "Usage", "Account Limit", "Available";
  		printf "%10s %9s + %14s %9s + %13s %9s\n",
  			"-"x10, "-"x9, "-"x14, "-"x9, "-"x13, "-"x9;
  	}
  }
--- 91,101 ----
  		printf "%-10s %9s | %14s %9s\n",
  			"User", "Usage", "Account", "Usage";
  		printf "%10s %9s + %14s %9s\n",
  			"-"x10, "-"x9, "-"x14, "-"x9;
  	} else {
! 		printf "%-10s %9s | %14s %9s | %13s %9s (SUs)\n",
  			"User", "Usage", "Account", "Usage", "Account Limit", "Available";
  		printf "%10s %9s + %14s %9s + %13s %9s\n",
  			"-"x10, "-"x9, "-"x14, "-"x9, "-"x13, "-"x9;
  	}
  }
***************
*** 444,454 ****
  
  # GrpCPUMins are not in 'sreport'
  while (<SACCTMGR>) {
  	# format is "acct_string|nnnn|" where nnnn is the number of GrpCPUMins allocated
  	if (/^([^|]+)\|([^|]*)/) {
! 		if ($2 ne "") {
  			$acc_limits{"\U$1"} = sprintf("%.0f", $2); # normalise account names to uppercase
  		} elsif (!exists($acc_limits{"\U$1"})) {
  			# store all accounts, even those without GrpCPUMins allocated, so we can report usage
  			$acc_limits{"\U$1"} = 0; # normalise account names to uppercase
  		}
--- 444,454 ----
  
  # GrpCPUMins are not in 'sreport'
  while (<SACCTMGR>) {
  	# format is "acct_string|nnnn|" where nnnn is the number of GrpCPUMins allocated
  	if (/^([^|]+)\|([^|]*)/) {
! 		if ($2 ne "" && !exists($acc_limits{"\U$1"})) {
  			$acc_limits{"\U$1"} = sprintf("%.0f", $2); # normalise account names to uppercase
  		} elsif (!exists($acc_limits{"\U$1"})) {
  			# store all accounts, even those without GrpCPUMins allocated, so we can report usage
  			$acc_limits{"\U$1"} = 0; # normalise account names to uppercase
  		}
